File: plugins/thoughtspot-to-sigma/skills/thoughtspot-to-sigma/scripts/pick_destination.py
#!/usr/bin/env python3
"""Shared destination picker for the *-to-sigma migration skills (Python port).
Produces the folderId where a migrated data model + workbook should land.
The SKILL drives the *asking*; this script lists candidates and creates folders.

  python3 pick_destination.py list
      -> {"workspaces":[{id,name}], "folders":[{id,name,parentId,parentName}],
          "myDocuments": "<id>"|null}
      Only EDIT-able folders are returned. folderId in a DM/workbook POST accepts
      a workspace id (lands in the workspace root) or a folder id.

  python3 pick_destination.py create --name "<NAME>" [--parent "<workspace-or-folder-id>"]
      -> {"id","name","parentId"}

Auth: SIGMA_API_TOKEN + SIGMA_BASE_URL if set (e.g. via get-token.sh); otherwise
minted from SIGMA_CLIENT_ID/SIGMA_CLIENT_SECRET (sourced from ~/.sigma-migration/env).
"""
import json, os, sys, urllib.request, urllib.parse, urllib.error

def _load_neutral_env():
    if os.environ.get("SIGMA_CLIENT_ID") or os.environ.get("SIGMA_API_TOKEN"):
        return
    p = os.path.expanduser("~/.sigma-migration/env")
    if os.path.exists(p):
        for line in open(p):
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            k, v = line.split("=", 1)
            os.environ.setdefault(k.strip(), v.strip().strip('"').strip("'"))

def _base():
    return os.environ.get("SIGMA_BASE_URL", "https://aws-api.sigmacomputing.com").rstrip("/")

def _token():
    tok = os.environ.get("SIGMA_API_TOKEN")
    if tok:
        return tok
    _load_neutral_env()
    data = urllib.parse.urlencode({
        "grant_type": "client_credentials",
        "client_id": os.environ["SIGMA_CLIENT_ID"],
        "client_secret": os.environ["SIGMA_CLIENT_SECRET"],
    }).encode()
    req = urllib.request.Request(_base() + "/v2/auth/token", data=data)
    return json.load(urllib.request.urlopen(req))["access_token"]

_TOK = None
def call(method, path, body=None):
    global _TOK
    if _TOK is None:
        _TOK = _token()
    url = _base() + path
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method,
        headers={"Authorization": "Bearer " + _TOK, "Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req) as r:
            raw = r.read().decode()
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        raise SystemExit(f"{method} {path} -> {e.code} {e.read().decode()[:300]}")

def my_documents_id():
    try:
        uid = (call("GET", "/v2/whoami") or {}).get("userId")
        if not uid:
            return None
        entries = (call("GET", f"/v2/members/{uid}/files?typeFilters=folder&limit=500") or {}).get("entries", [])
        for e in entries:
            if e.get("name") == "My Documents" or e.get("path") == "My Documents":
                return e.get("id")
    except Exception:
        return None
    return None

def cmd_list():
    ws = (call("GET", "/v2/workspaces?limit=500") or {}).get("entries", [])
    workspaces = [{"id": w.get("workspaceId") or w.get("id"), "name": w.get("name")} for w in ws]
    ws_name = {w["id"]: w["name"] for w in workspaces}
    fl = (call("GET", "/v2/files?typeFilters=folder&limit=500") or {}).get("entries", [])
    folders = [{"id": f["id"], "name": f["name"], "parentId": f.get("parentId"),
                "parentName": ws_name.get(f.get("parentId"))}
               for f in fl if f.get("permission") == "edit"]
    print(json.dumps({"workspaces": workspaces, "folders": folders,
                      "myDocuments": my_documents_id()}, indent=2))

def cmd_create(argv):
    name = parent = None
    i = 0
    while i < len(argv):
        if argv[i] == "--name":
            name = argv[i + 1]; i += 2
        elif argv[i] == "--parent":
            parent = argv[i + 1]; i += 2
        else:
            i += 1
    if not name:
        raise SystemExit("pick_destination create: --name is required")
    if not parent:
        parent = my_documents_id()
    body = {"type": "folder", "name": name}
    if parent:
        body["parentId"] = parent
    res = call("POST", "/v2/files", body)
    print(json.dumps({"id": res.get("id"), "name": res.get("name"), "parentId": res.get("parentId")}, indent=2))

if __name__ == "__main__":
    cmd = sys.argv[1] if len(sys.argv) > 1 else "list"
    if cmd == "list":
        cmd_list()
    elif cmd == "create":
        cmd_create(sys.argv[2:])
    else:
        raise SystemExit("usage: pick_destination.py [list | create --name NAME [--parent ID]]")

File: plugins/metabase-to-sigma/skills/metabase-to-sigma/scripts/pick-destination.mjs
#!/usr/bin/env node
// Shared destination picker for the *-to-sigma migration skills (Node port).
// Produces the folderId where a migrated data model + workbook should land.
// The SKILL drives the *asking*; this script lists candidates and creates folders.
//
//   node pick-destination.mjs list
//       -> { workspaces:[{id,name}], folders:[{id,name,parentId,parentName}], myDocuments }
//   node pick-destination.mjs create --name "<NAME>" [--parent "<workspace-or-folder-id>"]
//       -> { id, name, parentId }
//
// Auth: SIGMA_API_TOKEN + SIGMA_BASE_URL if set; else minted from
// SIGMA_CLIENT_ID/SIGMA_CLIENT_SECRET (sourced from ~/.sigma-migration/env).
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

function loadNeutralEnv() {
  if (process.env.SIGMA_CLIENT_ID || process.env.SIGMA_API_TOKEN) return;
  const p = path.join(os.homedir(), '.sigma-migration', 'env');
  if (!fs.existsSync(p)) return;
  for (const line of fs.readFileSync(p, 'utf8').split('\n')) {
    const t = line.trim();
    if (!t || t.startsWith('#') || !t.includes('=')) continue;
    const idx = t.indexOf('=');
    const k = t.slice(0, idx).trim();
    const v = t.slice(idx + 1).trim().replace(/^["']|["']$/g, '');
    if (process.env[k] === undefined) process.env[k] = v;
  }
}

const BASE = () => (process.env.SIGMA_BASE_URL || 'https://aws-api.sigmacomputing.com').replace(/\/$/, '');
let TOK = null;
async function token() {
  if (process.env.SIGMA_API_TOKEN) return process.env.SIGMA_API_TOKEN;
  loadNeutralEnv();
  const body = new URLSearchParams({ grant_type: 'client_credentials',
    client_id: process.env.SIGMA_CLIENT_ID, client_secret: process.env.SIGMA_CLIENT_SECRET });
  const r = await fetch(BASE() + '/v2/auth/token', { method: 'POST', body });
  return (await r.json()).access_token;
}
async function call(method, p, body) {
  if (!TOK) TOK = await token();
  const r = await fetch(BASE() + p, { method,
    headers: { Authorization: 'Bearer ' + TOK, 'Content-Type': 'application/json' },
    body: body !== undefined ? JSON.stringify(body) : undefined });
  const raw = await r.text();
  if (!r.ok) { console.error(`${method} ${p} -> ${r.status} ${raw.slice(0, 300)}`); process.exit(1); }
  return raw ? JSON.parse(raw) : {};
}
async function myDocumentsId() {
  try {
    const uid = (await call('GET', '/v2/whoami'))?.userId;
    if (!uid) return null;
    const entries = (await call('GET', `/v2/members/${uid}/files?typeFilters=folder&limit=500`))?.entries || [];
    const hit = entries.find(e => e.name === 'My Documents' || e.path === 'My Documents');
    return hit ? hit.id : null;
  } catch { return null; }
}
async function cmdList() {
  const ws = ((await call('GET', '/v2/workspaces?limit=500')).entries || [])
    .map(w => ({ id: w.workspaceId || w.id, name: w.name }));
  const wsName = Object.fromEntries(ws.map(w => [w.id, w.name]));
  const folders = ((await call('GET', '/v2/files?typeFilters=folder&limit=500')).entries || [])
    .filter(f => f.permission === 'edit')
    .map(f => ({ id: f.id, name: f.name, parentId: f.parentId, parentName: wsName[f.parentId] || null }));
  console.log(JSON.stringify({ workspaces: ws, folders, myDocuments: await myDocumentsId() }, null, 2));
}
async function cmdCreate(argv) {
  let name = null, parent = null;
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--name') { name = argv[++i]; }
    else if (argv[i] === '--parent') { parent = argv[++i]; }
  }
  if (!name) { console.error('pick-destination create: --name is required'); process.exit(1); }
  if (!parent) parent = await myDocumentsId();
  const body = { type: 'folder', name };
  if (parent) body.parentId = parent;
  const res = await call('POST', '/v2/files', body);
  console.log(JSON.stringify({ id: res.id, name: res.name, parentId: res.parentId }, null, 2));
}
const cmd = process.argv[2] || 'list';
if (cmd === 'list') await cmdList();
else if (cmd === 'create') await cmdCreate(process.argv.slice(3));
else { console.error('usage: pick-destination.mjs [list | create --name NAME [--parent ID]]'); process.exit(1); }
